server/util/perms/perms.go

```diff
@@ -255,3 +255,30 @@ func AuthenticateSelectedGroupID(ctx context.Context, env environment.Env, proto
 	}
 	return groupID, nil
 }
+
+// AuthenticatedGroupID returns the authenticated group ID from the given
+// context. This is preferred for API requests, since the group ID can be
+// determined directly from the API key. UI requests should instead use
+// `AuthenticateSelectedGroupID`, since the API key is not available, and the
+// user's selected group ID needs to be taken into account.
+func AuthenticatedGroupID(ctx context.Context, env environment.Env) (string, error) {
+	u, err := AuthenticatedUser(ctx, env)
+	if err != nil {
+		return "", err
+	}
+	groupID := u.GetGroupID()
+	if groupID == "" {
+		return "", status.FailedPreconditionError("Authenticated user does not have an associated group ID")
+	}
+	return groupID, nil
+}
+
+// IsAnonymousUserError can be used to check whether an error returned by
+// functions which return the authenticated user (such as AuthenticatedUser or
+// AuthenticateSelectedGroupID) is due to an anonymous user accessing the
+// service. This is useful for allowing anonymous users to proceed, in cases
+// where anonymous usage is explicitly enabled in the app config, and we support
+// anonymous usage for the part of the service where this is used.
+func IsAnonymousUserError(err error) bool {
+	return status.IsUnauthenticatedError(err) || status.IsPermissionDeniedError(err) || status.IsUnimplementedError(err)
+}
```

Comment:
This feels really broad (and potentially dangerous).-- what do the current callsites that permit anonymous usage look like now?

Reply:
We do this in a few other places --

![Screenshot from 2021-09-16 13-24-52](https://user-images.githubusercontent.com/2414826/133657573-352050ee-6dd8-4267-af99-01d46b28219f.png)

Reply:
Resolved offline to update these callsites in a follow-up PR.